widgyts/dataset_viewer.py

```diff
@@ -0,0 +1,252 @@
+import numpy as np
+import json
+import pythreejs
+import traitlets
+import ipywidgets
+from IPython.display import display, JSON
+import matplotlib.cm as mcm
+import matplotlib.colors as mcolors
+from yt.data_objects.api import Dataset
+from yt.units.yt_array import display_ytarray
+
+
+class DatasetViewerComponent(traitlets.HasTraits):
+    ds = traitlets.Instance(Dataset)
+    viewer = traitlets.ForwardDeclaredInstance("DatasetViewer")
+
+
+class DatasetViewer(traitlets.HasTraits):
+    ds = traitlets.Instance(Dataset)
+    components = traitlets.List(trait=traitlets.Instance(DatasetViewerComponent))
+
+    @traitlets.default("components")
+    def _components_default(self):
+        adv = AMRDomainViewer(ds=self.ds, viewer=self)
+        fdv = FieldDefinitionViewer(ds=self.ds, viewer=self)
+        pv = ParametersViewer(ds=self.ds, viewer=self)
+        return [adv, fdv, pv]
+
+    def widget(self):
+        tab = ipywidgets.Tab(
+            children=[_.widget() for _ in self.components],
+            layout=ipywidgets.Layout(height="500px"),
+        )
+        for i, c in enumerate(self.components):
+            tab.set_title(i, c.name)
+        return tab
+
+
+class FieldDefinitionViewer(DatasetViewerComponent):
+    name = "Fields"
+
+    def widget(self):
+        out = ipywidgets.Output()
+        with out:
+            display(self.ds.fields)
+        return out
+
+
+class DomainViewer(DatasetViewerComponent):
+    domain_axes = traitlets.Instance(pythreejs.AxesHelper)
+    name = "Domain Viewer"
+
+    @traitlets.default("domain_axes")
+    def _domain_axes_default(self):
+        offset_vector = (self.ds.domain_left_edge - self.ds.domain_center) * 0.1
+        position = tuple(
+            (self.ds.domain_left_edge + offset_vector).in_units("code_length").d
+        )
+        # We probably don't want to use the AxesHelper as it doesn't expose the
+        # material, which can result in it not being easy to see.  But for now...
+        ah = pythreejs.AxesHelper(
+            position=position,
+            scale=tuple(self.ds.domain_width.in_units("code_length").d),
+        )
+        return ah
+
+
+_CORNER_INDICES = np.array(
+    [0, 1, 1, 2, 2, 3, 3, 0, 4, 5, 5, 6, 6, 7, 7, 4, 0, 4, 1, 5, 2, 6, 3, 7],
+    dtype="uint32",
+)
+
+
+class AMRDomainViewer(DomainViewer):
+    grid_views = traitlets.List(trait=traitlets.Instance(pythreejs.LineSegments))
+    renderer = traitlets.Instance(pythreejs.Renderer)
+    r2_falloff = traitlets.Instance(pythreejs.Texture)
+    colormap_texture = traitlets.Instance(pythreejs.Texture)
+
+    @traitlets.default("grid_views")
+    def _grid_views_default(self):
+        # This needs to generate the geometries and access the materials
+        grid_views = []
+        cmap = mcm.get_cmap("inferno")
+        for level in range(self.ds.max_level + 1):
+            # We truncate at half of the colormap so that we just get a slight
```

Comment:
Maybe we should make this an attribute so users can play around with the progression (and set the default to half)?

Reply:
Sure, sounds good